File: DynamicTypes/implementations/cpp-classic/numberWords/numberWords.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <cmath>
#include "MiniscriptTypes.h"
#include "SplitJoin.h"

using namespace MiniScript;

// Global word arrays - equivalent to MiniScript's split arrays
StringList singles;
StringList teens;
StringList tys;
StringList ions;

void initializeWordArrays() {
    // singles = " one two three four five six seven eight nine ".split
    singles = Split(String(" one two three four five six seven eight nine "), String(" "));
    
    // teens = "ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen ".split
    teens = Split(String("ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen "), String(" "));
    
    // tys = "  twenty thirty forty fifty sixty seventy eighty ninety".split
    tys = Split(String("  twenty thirty forty fifty sixty seventy eighty ninety"), String(" "));
    
    // ions = "thousand million billion".split
    ions = Split(String("thousand million billion"), String(" "));
}

String numberToText(long n) {
    if (n == 0) return String("zero");
    
    long a = std::abs(n);
    String r = String("");  // result
    
    // Process each scale (units, thousands, millions, billions)
    for (long ionIndex = 0; ionIndex < (long)ions.Count(); ionIndex++) {
        String u = ions[ionIndex];
        
        long h = a % 100;
        if (h > 0 && h < 10) {
            r = singles[h] + String(" ") + r;
        }
        if (h > 9 && h < 20) {
            r = teens[h-10] + String(" ") + r;
        }
        if (h > 19 && h < 100) {
            String hyphen = (h % 10 > 0) ? String("-") : String("");
            r = tys[h/10] + hyphen + singles[h%10] + String(" ") + r;
        }
        
        h = (a % 1000) / 100;
        if (h) {
            r = singles[h] + String(" hundred ") + r;
        }
        
        a = a / 1000;
        if (a == 0) break;
        if (a % 1000 > 0) {
            r = u + String(" ") + r;
        }
    }
    
    if (n < 0) {
        r = String("negative ") + r;
    }
    
    return r.Trim();
}

long textToNumber(String s) {
    if (s == String("zero")) return 0;
    
    // Replace hyphens with spaces and split
    String cleanS = s.Replace("-", " ");
    StringList words = Split(cleanS, String(" "));
    
    long result = 0;
    long ionVal = 0;
    bool negative = false;
    long maxi = words.Count() - 1;
    long i = 0;
    
    while (i <= maxi) {
        String word = words[i];
        
        if (word == String("negative")) {
            negative = true;
            i++;
            continue;
        }
        
        // Check for scale words (thousand, million, billion)
        long idx = -1;
        for (long j = 0; j < (long)ions.Count(); j++) {
            if (ions[j] == word) {
                idx = j;
                break;
            }
        }
        
        if (idx != -1) {
            long multipliers[] = {1000, 1000000, 1000000000};
            result += ionVal * multipliers[idx];
            ionVal = 0;
            i++;
            continue;
        }
        
        long wordVal = 0;
        
        // Check singles
        idx = -1;
        for (long j = 0; j < (long)singles.Count(); j++) {
            if (singles[j] == word) {
                idx = j;
                break;
            }
        }
        
        if (idx != -1) {
            wordVal = idx;
        } else {
            // Check tys (twenties, thirties, etc.)
            idx = -1;
            for (long j = 0; j < (long)tys.Count(); j++) {
                if (tys[j] == word) {
                    idx = j;
                    break;
                }
            }
            
            if (idx != -1) {
                wordVal = idx * 10;
            } else {
                // Check teens
                idx = -1;
                for (long j = 0; j < (long)teens.Count(); j++) {
                    if (teens[j] == word) {
                        idx = j;
                        break;
                    }
                }
                
                if (idx != -1) {
                    wordVal = idx + 10;
                } else {
                    std::cout << "Unexpected word: " << word.c_str() << std::endl;
                    return 0;
                }
            }
        }
        
        // Handle "hundred"
        if (i < maxi && words[i+1] == String("hundred")) {
            wordVal *= 100;
            i++;
        }
        
        ionVal += wordVal;
        i++;
    }
    
    result += ionVal;
    if (negative) result = -result;
    
    return result;
}

double get_time() {
    auto now = std::chrono::high_resolution_clock::now();
    auto duration = now.time_since_epoch();
    auto nanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(duration);
    return nanoseconds.count() / 1000000000.0;
}

void runBenchmark(long n = 10000) {
    double t0 = get_time();
    
    for (long i = 0; i < n; i++) {
        String s = numberToText(i);
        long i2 = textToNumber(s);
        if (i2 != i) {
            std::cout << "Oops! Failed on " << i << ":" << std::endl;
            std::cout << s.c_str() << " --> " << i2 << std::endl;
        }
    }
    
    double t1 = get_time();
    double elapsed = t1 - t0;
    
    std::cout << "numberWords(" << n << ") time: " << elapsed << " seconds" << std::endl;
}

void runCorrectnessTests() {
    std::cout << "Correctness checks:" << std::endl;
    
    long testNumbers[] = {-1234, 0, 7, 42, 4325, 1000004, 214837564};
    int numTests = sizeof(testNumbers) / sizeof(testNumbers[0]);
    
    for (int i = 0; i < numTests; i++) {
        long n = testNumbers[i];
        String words = numberToText(n);
        long backToNum = textToNumber(words);
        
        std::cout << n << ": " << words.c_str() << " -> " << backToNum;
        if (backToNum != n) {
            std::cout << " ERROR --^" << std::endl;
            return;
        } else {
            std::cout << std::endl;
        }
    }
}

int main() {
    std::cout << "MiniScript::Value NumberWords Benchmark" << std::endl;
    std::cout << "=======================================" << std::endl;
    
    initializeWordArrays();
    
    runCorrectnessTests();
    std::cout << std::endl;
    
    runBenchmark();
    
    return 0;
}

File: DynamicTypes/implementations/cs-nan-box/NumberWords.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace ScriptingVM.Benchmarks
{
    class NumberWords
    {
        // Global word arrays - equivalent to MiniScript's split arrays
        static string[] singles = { "", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        static string[] teens = { "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen" };
        static string[] tys = { "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety" };
        static string[] ions = { "thousand", "million", "billion" };
        
        static Value NumberToText(long n)
        {
            if (n == 0)
                return Value.FromString("zero");
            
            long a = Math.Abs(n);
            StringBuilder r = new StringBuilder();
            
            // Process each scale (units, thousands, millions, billions)
            for (int ionIndex = 0; ionIndex < ions.Length; ionIndex++)
            {
                string u = ions[ionIndex];
                
                long h = a % 100;
                if (h > 0 && h < 10)
                {
                    r.Insert(0, singles[h] + " ");
                }
                if (h > 9 && h < 20)
                {
                    r.Insert(0, teens[h - 10] + " ");
                }
                if (h > 19 && h < 100)
                {
                    string hyphen = (h % 10 > 0) ? "-" : "";
                    r.Insert(0, tys[h / 10] + hyphen + singles[h % 10] + " ");
                }
                
                h = (a % 1000) / 100;
                if (h > 0)
                {
                    r.Insert(0, singles[h] + " hundred ");
                }
                
                a = a / 1000;
                if (a == 0) break;
                if (a % 1000 > 0)
                {
                    r.Insert(0, u + " ");
                }
            }
            
            if (n < 0)
            {
                r.Insert(0, "negative ");
            }
            
            // Trim the result
            string result = r.ToString().Trim();
            return Value.FromString(result);
        }
        
        static long TextToNumber(Value s_val)
        {
            if (!s_val.IsString)
                return 0;
                
            string s = s_val.ToString();
            
            if (s == "zero")
                return 0;
            
            // Replace hyphens with spaces and split
            string cleanS = s.Replace("-", " ");
            string[] words = cleanS.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
            
            long result = 0;
            long ionVal = 0;
            bool negative = false;
            
            for (int i = 0; i < words.Length; i++)
            {
                string word = words[i];
                
                if (word == "negative")
                {
                    negative = true;
                    continue;
                }
                
                // Check for scale words
                int idx = Array.IndexOf(ions, word);
                if (idx != -1)
                {
                    long[] multipliers = { 1000, 1000000, 1000000000 };
                    result += ionVal * multipliers[idx];
                    ionVal = 0;
                    continue;
                }
                
                long wordVal = 0;
                
                // Check singles
                idx = Array.IndexOf(singles, word);
                if (idx != -1)
                {
                    wordVal = idx;
                }
                else
                {
                    // Check tys
                    idx = Array.IndexOf(tys, word);
                    if (idx != -1)
                    {
                        wordVal = idx * 10;
                    }
                    else
                    {
                        // Check teens
                        idx = Array.IndexOf(teens, word);
                        if (idx != -1)
                        {
                            wordVal = idx + 10;
                        }
                        else
                        {
                            Console.WriteLine($"Unexpected word: {word}");
                            return 0;
                        }
                    }
                }
                
                // Handle "hundred"
                if (i < words.Length - 1 && words[i + 1] == "hundred")
                {
                    wordVal *= 100;
                    i++; // Skip "hundred"
                }
                
                ionVal += wordVal;
            }
            
            result += ionVal;
            if (negative) result = -result;
            
            return result;
        }
        
        static void RunBenchmark(long n)
        {
            var stopwatch = Stopwatch.StartNew();
            
            for (long i = 0; i < n; i++)
            {
                var s = NumberToText(i);
                long i2 = TextToNumber(s);
                if (i2 != i)
                {
                    Console.WriteLine($"Oops! Failed on {i}:");
                    Console.WriteLine($"'{s}' --> {i2}");
                    break;
                }
            }
            
            stopwatch.Stop();
            Console.WriteLine($"numberWords({n}) time: {stopwatch.ElapsedMilliseconds / 1000.0:F3} seconds");
        }
        
        static void RunCorrectnessTests()
        {
            Console.WriteLine("Correctness checks:");
            
            long[] testNumbers = { -1234, 0, 7, 42, 4325, 1000004, 214837564 };
            
            foreach (long n in testNumbers)
            {
                var words = NumberToText(n);
                long backToNum = TextToNumber(words);
                
                Console.Write($"{n}: {words} -> {backToNum}");
                if (backToNum != n)
                {
                    Console.WriteLine(" ERROR --^");
                    return;
                }
                Console.WriteLine();
            }
        }
        
        static void Main(string[] args)
        {
            Console.WriteLine("C# NaN Boxing NumberWords Benchmark");
            Console.WriteLine("====================================");
            Console.WriteLine();
            
            Console.WriteLine($"Before tests: {HandlePool.GetCount()} objects in HandlePool");
            
            RunCorrectnessTests();
            Console.WriteLine($"After correctness tests: {HandlePool.GetCount()} objects in HandlePool");
            Console.WriteLine();
            
            RunBenchmark(10000);
            
            Console.WriteLine($"After benchmark: {HandlePool.GetCount()} objects in HandlePool");
        }
    }
}